### PR TITLE
Optimize naive top-k masking in fused router

### PR DESCRIPTION
**Pull request by yosh20004**
# Description

Refactor the fused router's naive top-k selection path to track already-selected entries with a per-lane mask

Fixes # (issue)

## Type of change

- [ ] Documentation change (change only to the documentation, either a fix or a new content)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Infra/Build change
- [x] Code refactoring

## Changes

- Refactor `naive_topk_and_mask` in `transformer_engine/common/fused_router/utils.h` to track per-lane selections with a local bitmask
- Preserve the existing function interface and current fused router call sites

# Checklist:

- [x] I have read and followed the [contributing guidelines](https://github.com/NVIDIA/TransformerEngine/blob/main/CONTRIBUTING.rst)
- [x] The functionality is complete
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

### CODE REVIEW

**Comment by yosh20004** [comment]
### Performance test script
```python
# Copyright (c) 2022-2026, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
#
# See LICENSE for license information.

import os
from typing import Any, Callable, Optional, Tuple, cast

import pytest
import torch

from transformer_engine.pytorch.router import (
    fused_compute_score_for_moe_aux_loss,
    fused_moe_aux_loss,
    fused_topk_with_score_function,
)


seed = 42
torch.manual_seed(seed)
if torch.cuda.is_available():
    torch.cuda.manual_seed(seed)


def _cuda() -> Any:
    return getattr(torch, "cuda", None)


def _require_perf_env() -> None:
    if _cuda() is None or not _cuda().is_available():
        pytest.skip("CUDA is not available.")
    if os.getenv("TE_RUN_PERF_TESTS", "1") != "1":
        pytest.skip("Set TE_RUN_PERF_TESTS=1 to enable router performance tests.")


def _benchmark_cuda_kernel(
    fn: Callable[[], object], warmup: int = 500, iters: int = 2000
) -> float:
    start_event = _cuda().Event(enable_timing=True)
    end_event = _cuda().Event(enable_timing=True)

    for _ in range(warmup):
        fn()
    _cuda().synchronize()

    start_event.record()
    for _ in range(iters):
        fn()
    end_event.record()
    _cuda().synchronize()

    return start_event.elapsed_time(end_event) / iters


def group_limited_topk(
    scores: torch.Tensor,
    topk: int,
    num_tokens: int,
    num_experts: int,
    num_groups: int,
    group_topk: int,
) -> Tuple[torch.Tensor, torch.Tensor]:
    group_scores = (
        scores.view(num_tokens, num_groups, -1)
        .topk(topk // group_topk, dim=-1)[0]
        .sum(dim=-1)
    )
    group_idx = torch.topk(group_scores, k=group_topk, dim=-1, sorted=False)[1]
    group_mask = torch.zeros_like(group_scores)
    group_mask.scatter_(1, group_idx, 1)

    score_mask = (
        group_mask.unsqueeze(-1)
        .expand(num_tokens, num_groups, num_experts // num_groups)
        .reshape(num_tokens, -1)
    )
    masked_scores = scores.masked_fill(~score_mask.bool(), float("-inf"))
    probs, top_indices = torch.topk(masked_scores, k=topk, dim=-1)
    return probs, top_indices


def topk_score_function_pytorch(
    logits: torch.Tensor,
    topk: int,
    use_pre_softmax: bool = False,
    num_groups: Optional[int] = None,
    group_topk: Optional[int] = None,
    scaling_factor: Optional[float] = None,
    score_function: str = "softmax",
    expert_bias: Optional[torch.Tensor] = None,
) -> Tuple[torch.Tensor, torch.Tensor]:
    num_tokens, num_experts = logits.shape

    def compute_topk(
        scores: torch.Tensor,
        topk_value: int,
        num_groups_value: Optional[int] = None,
        group_topk_value: Optional[int] = None,
    ) -> Tuple[torch.Tensor, torch.Tensor]:
        if group_topk_value:
            assert num_groups_value is not None
            return group_limited_topk(
                scores=scores,
                topk=topk_value,
                num_tokens=num_tokens,
                num_experts=num_experts,
                num_groups=num_groups_value,
                group_topk=group_topk_value,
            )
        return torch.topk(scores, k=topk_value, dim=1)

    if score_function == "softmax":
        if use_pre_softmax:
            scores = torch.softmax(logits, dim=-1, dtype=torch.float32).type_as(logits)
            probs, top_indices = compute_topk(scores, topk, num_groups, group_topk)
        else:
            scores, top_indices = compute_topk(logits, topk, num_groups, group_topk)
            probs = torch.softmax(scores, dim=-1, dtype=torch.float32).type_as(logits)
    elif score_function in ("sigmoid", "sqrtsoftplus"):
        if score_function == "sigmoid":
            scores = torch.sigmoid(logits.float()).type_as(logits)
        else:
            scores = torch.nn.functional.softplus(logits.float()).sqrt().type_as(logits)
        if expert_bias is not None:
            scores_for_routing = scores + expert_bias
            _, top_indices = compute_topk(
                scores_for_routing, topk, num_groups, group_topk
            )
            scores = torch.gather(scores, dim=1, index=top_indices).type_as(logits)
        else:
            scores, top_indices = compute_topk(scores, topk, num_groups, group_topk)
        probs = (
            scores / (scores.sum(dim=-1, keepdim=True) + 1e-20) if topk > 1 else scores
        )
    else:
        raise ValueError(f"Invalid score_function: {score_function}")

    if scaling_factor:
        probs = probs * scaling_factor

    topk_masked_gates = torch.zeros_like(logits).scatter(1, top_indices, probs)
    topk_map = torch.zeros_like(logits).int().scatter(1, top_indices, 1).bool()
    return topk_masked_gates, topk_map


def compute_scores_for_aux_loss_pytorch(
    logits: torch.Tensor, topk: int, score_function: str
) -> Tuple[torch.Tensor, torch.Tensor]:
    if score_function == "softmax":
        scores = torch.softmax(logits, dim=-1, dtype=torch.float32)
    elif score_function == "sigmoid":
        scores = torch.sigmoid(logits.float())
        scores = scores / (scores.sum(dim=-1, keepdim=True) + 1e-20)
    elif score_function == "sqrtsoftplus":
        scores = torch.nn.functional.softplus(logits.float()).sqrt()
        scores = scores / (scores.sum(dim=-1, keepdim=True) + 1e-20)
    else:
        raise ValueError(f"Invalid score_function: {score_function}")

    _, top_indices = torch.topk(scores, k=topk, dim=1)
    routing_map = torch.zeros_like(logits).int().scatter(1, top_indices, 1).bool()
    return routing_map, scores


def aux_loss_pytorch(
    probs: torch.Tensor,
    tokens_per_expert: torch.Tensor,
    total_num_tokens: int,
    topk: int,
    num_experts: int,
    moe_aux_loss_coeff: float,
) -> torch.Tensor:
    aggregated_probs_per_expert = probs.sum(dim=0)
    return torch.sum(aggregated_probs_per_expert * tokens_per_expert) * (
        num_experts * moe_aux_loss_coeff / (topk * total_num_tokens * total_num_tokens)
    )


def _make_router_logits(
    dtype: torch.dtype, num_tokens: int, num_experts: int, score_function: str
) -> torch.Tensor:
    if score_function in ("sigmoid", "sqrtsoftplus"):
        offset = (
            torch.arange(-num_tokens // 2, num_tokens // 2, dtype=dtype, device="cuda")
            * 1e-4
        )
        logits = (
            torch.arange(
                -num_experts // 2, num_experts // 2, device="cuda", dtype=dtype
            )
            * 1e-2
        )
        return logits.unsqueeze(0).repeat(num_tokens, 1) + offset.unsqueeze(1)

    logits = (
        torch.arange(
            -num_tokens * num_experts // 2,
            num_tokens * num_experts // 2,
            device="cuda",
            dtype=dtype,
        )
        * 1e-4
    )
    return logits.view(num_tokens, num_experts)


def _make_router_bias(num_experts: int, dtype: torch.dtype) -> torch.Tensor:
    bias = torch.arange(num_experts, device="cuda", dtype=dtype) * 0.1
    return torch.flip(bias, dims=[0])


def _make_aux_loss_probs(
    dtype: torch.dtype, num_tokens: int, num_experts: int
) -> torch.Tensor:
    offset = (
        torch.arange(-num_tokens // 2, num_tokens // 2, dtype=dtype, device="cuda")
        * 1e-4
    )
    probs = (
        torch.arange(-num_experts // 2, num_experts // 2, device="cuda", dtype=dtype)
        * 1e-2
    )
    probs = probs.unsqueeze(0).repeat(num_tokens, 1) + offset.unsqueeze(1)
    return probs.view(num_tokens, num_experts)


def _fused_expert_bias(expert_bias: Optional[torch.Tensor]) -> torch.Tensor:
    if expert_bias is None:
        return cast(torch.Tensor, None)
    return expert_bias


def _print_perf_result(case_name: str, torch_ms: float, fused_ms: float) -> None:
    speedup = torch_ms / fused_ms
    print(
        f"{case_name}: torch={torch_ms:.6f} ms, fused={fused_ms:.6f} ms, speedup={speedup:.4f}x"
    )


def _perf_assert_message(case_name: str, torch_ms: float, fused_ms: float) -> str:
    return (
        f"{case_name} perf result: torch={torch_ms:.6f} ms, "
        f"fused={fused_ms:.6f} ms, speedup={torch_ms / fused_ms:.4f}x"
    )


@pytest.mark.parametrize("dtype", [torch.float32])
@pytest.mark.parametrize(
    "num_tokens,num_experts,topk",
    [(8192, 256, 8), (16384, 256, 8)],
    ids=["large", "xlarge"],
)
@pytest.mark.parametrize(
    "score_function,use_pre_softmax,enable_bias,num_groups,group_topk,scaling_factor",
    [
        ("softmax", True, False, None, None, None),
        ("softmax", False, False, 8, 4, 1.2),
        ("sigmoid", False, False, None, None, None),
        ("sigmoid", False, True, 8, 4, 1.2),
        ("sqrtsoftplus", False, False, None, None, None),
        ("sqrtsoftplus", False, True, 8, 4, 1.2),
    ],
    ids=[
        "softmax_pre",
        "softmax_grouped",
        "sigmoid_plain",
        "sigmoid_grouped_bias",
        "sqrtsoftplus_plain",
        "sqrtsoftplus_grouped_bias",
    ],
)
def test_fused_topk_router_perf_against_torch(
    dtype,
    num_tokens,
    num_experts,
    topk,
    score_function,
    use_pre_softmax,
    enable_bias,
    num_groups,
    group_topk,
    scaling_factor,
    record_property,
):
    _require_perf_env()

    logits = _make_router_logits(dtype, num_tokens, num_experts, score_function)
    expert_bias = None
    if enable_bias and score_function in ("sigmoid", "sqrtsoftplus"):
        expert_bias = _make_router_bias(num_experts, dtype)

    torch_probs, torch_map = topk_score_function_pytorch(
        logits=logits,
        topk=topk,
        use_pre_softmax=use_pre_softmax,
        num_groups=num_groups,
        group_topk=group_topk,
        scaling_factor=scaling_factor,
        score_function=score_function,
        expert_bias=expert_bias,
    )
    fused_probs, fused_map = cast(
        Tuple[torch.Tensor, torch.Tensor],
        fused_topk_with_score_function(
            logits=logits,
            topk=topk,
            use_pre_softmax=use_pre_softmax,
            num_groups=num_groups,
            group_topk=group_topk,
            scaling_factor=scaling_factor,
            score_function=score_function,
            expert_bias=_fused_expert_bias(expert_bias),
        ),
    )

    torch_ms = _benchmark_cuda_kernel(
        lambda: topk_score_function_pytorch(
            logits=logits,
            topk=topk,
            use_pre_softmax=use_pre_softmax,
            num_groups=num_groups,
            group_topk=group_topk,
            scaling_factor=scaling_factor,
            score_function=score_function,
            expert_bias=expert_bias,
        )
    )
    fused_ms = _benchmark_cuda_kernel(
        lambda: fused_topk_with_score_function(
            logits=logits,
            topk=topk,
            use_pre_softmax=use_pre_softmax,
            num_groups=num_groups,
            group_topk=group_topk,
            scaling_factor=scaling_factor,
            score_function=score_function,
            expert_bias=_fused_expert_bias(expert_bias),
        )
    )

    case_name = (
        f"topk_router[{score_function}]"
        f"[tokens={num_tokens},experts={num_experts},topk={topk}]"
        f"[pre={use_pre_softmax},groups={group_topk is not None},bias={enable_bias}]"
    )
    record_property("torch_ms", round(torch_ms, 6))
    record_property("fused_ms", round(fused_ms, 6))
    record_property("speedup", round(torch_ms / fused_ms, 6))
    _print_perf_result(case_name, torch_ms, fused_ms)

    assert torch_ms > 0, _perf_assert_message(case_name, torch_ms, fused_ms)
    assert fused_ms > 0, _perf_assert_message(case_name, torch_ms, fused_ms)
    torch.testing.assert_close(torch_probs, fused_probs)
    torch.testing.assert_close(torch_map, fused_map)


@pytest.mark.parametrize("dtype", [torch.float32])
@pytest.mark.parametrize(
    "num_tokens,num_experts,topk",
    [(8192, 256, 8), (16384, 256, 8)],
    ids=["large", "xlarge"],
)
@pytest.mark.parametrize("score_function", ["softmax", "sigmoid", "sqrtsoftplus"])
def test_fused_scores_for_aux_loss_perf_against_torch(
    dtype, num_tokens, num_experts, topk, score_function, record_property
):
    _require_perf_env()

    logits = _make_router_logits(dtype, num_tokens, num_experts, score_function)

    torch_map, torch_scores = compute_scores_for_aux_loss_pytorch(
        logits=logits,
        topk=topk,
        score_function=score_function,
    )
    fused_map, fused_scores = cast(
        Tuple[torch.Tensor, torch.Tensor],
        fused_compute_score_for_moe_aux_loss(
            logits=logits,
            topk=topk,
            score_function=score_function,
        ),
    )

    torch_ms = _benchmark_cuda_kernel(
        lambda: compute_scores_for_aux_loss_pytorch(
            logits=logits,
            topk=topk,
            score_function=score_function,
        )
    )
    fused_ms = _benchmark_cuda_kernel(
        lambda: fused_compute_score_for_moe_aux_loss(
            logits=logits,
            topk=topk,
            score_function=score_function,
        )
    )

    case_name = (
        f"scores_for_aux_loss[{score_function}]"
        f"[tokens={num_tokens},experts={num_experts},topk={topk}]"
    )
    record_property("torch_ms", round(torch_ms, 6))
    record_property("fused_ms", round(fused_ms, 6))
    record_property("speedup", round(torch_ms / fused_ms, 6))
    _print_perf_result(case_name, torch_ms, fused_ms)

    assert torch_ms > 0, _perf_assert_message(case_name, torch_ms, fused_ms)
    assert fused_ms > 0, _perf_assert_message(case_name, torch_ms, fused_ms)
    torch.testing.assert_close(torch_scores, fused_scores)
    torch.testing.assert_close(torch_map, fused_map)


@pytest.mark.parametrize("dtype", [torch.float32])
@pytest.mark.parametrize(
    "num_tokens,num_experts,topk",
    [(14234, 256, 4), (28672, 256, 8)],
    ids=["large", "xlarge"],
)
@pytest.mark.parametrize("coeff", [0.01, 0.05])
def test_fused_moe_aux_loss_perf_against_torch(
    dtype, num_tokens, num_experts, topk, coeff, record_property
):
    _require_perf_env()

    probs = _make_aux_loss_probs(dtype, num_tokens, num_experts)
    tokens_per_expert = torch.randint(
        1, 1000, (num_experts,), device="cuda", dtype=torch.int32
    )

    torch_loss = aux_loss_pytorch(
        probs=probs,
        tokens_per_expert=tokens_per_expert,
        total_num_tokens=num_tokens,
        topk=topk,
        num_experts=num_experts,
        moe_aux_loss_coeff=coeff,
    )
    fused_loss = fused_moe_aux_loss(
        probs=probs,
        tokens_per_expert=tokens_per_expert,
        total_num_tokens=num_tokens,
        num_experts=num_experts,
        topk=topk,
        coeff=coeff,
    )

    torch_ms = _benchmark_cuda_kernel(
        lambda: aux_loss_pytorch(
            probs=probs,
            tokens_per_expert=tokens_per_expert,
            total_num_tokens=num_tokens,
            topk=topk,
            num_experts=num_experts,
            moe_aux_loss_coeff=coeff,
        )
    )
    fused_ms = _benchmark_cuda_kernel(
        lambda: fused_moe_aux_loss(
            probs=probs,
            tokens_per_expert=tokens_per_expert,
            total_num_tokens=num_tokens,
            num_experts=num_experts,
            topk=topk,
            coeff=coeff,
        )
    )

    case_name = f"moe_aux_loss[tokens={num_tokens},experts={num_experts},coeff={coeff}]"
    record_property("torch_ms", round(torch_ms, 6))
    record_property("fused_ms", round(fused_ms, 6))
    record_property("speedup", round(torch_ms / fused_ms, 6))
    _print_perf_result(case_name, torch_ms, fused_ms)

    assert torch_ms > 0, _perf_assert_message(case_name, torch_ms, fused_ms)
    assert fused_ms > 0, _perf_assert_message(case_name, torch_ms, fused_ms)
    torch.testing.assert_close(torch_loss, fused_loss)

```

### Result (H200 cuda13.1)
```
(1) (test_topk_router[softmax][tokens=8192,experts=256,topk=8][pre=True,groups=False,bias=False])
    before: 0.044309 ms
    after: 0.025076 ms
    speedup: 1.77x

(2) (test_topk_router[sigmoid][tokens=16384,experts=256,topk=8][pre=False,groups=False,bias=False])
    before: 0.077351 ms
    after: 0.042781 ms
    speedup: 1.81x

(3) (test_topk_router[softmax][tokens=16384,experts=256,topk=8][pre=False,groups=True,bias=False])
    before: 0.112017 ms
    after: 0.081958 ms
    speedup: 1.37x

(4) (test_scores_for_aux_loss[sigmoid][tokens=16384,experts=256,topk=8])
    before: 0.083227 ms
    after: 0.045734 ms
    speedup: 1.82x

(5) (test_topk_router[sqrtsoftplus][tokens=16384,experts=256,topk=8][pre=False,groups=True,bias=True])
    before: 0.119349 ms
    after: 0.089024 ms
    speedup: 1.34x
```

**Comment by denera** [comment]
Closing this because of no action on the correctness fixes for quite a while. Please re-open if you'd like to re-visit it with fixes.
